File: 5-longest-palindromic-substring/longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        if n < 0:
            return 0
        sol = 1
        sol_values = s[0]
        mem = [[-1] * n for _ in range(n)]

        for i in range(n):
            mem[i][i] = 1

        for i in range(1, n):
            for j in range(i, n):
                k = j - i
                if s[j] != s[k]:
                    mem[j][k] = -1 
                    continue
                if j-1 < k + 1:
                    mem[j][k] = 2
                elif (mem[j-1][k+1] != -1):
                    mem[j][k] = max(mem[j][k], 2 + mem[j-1][k+1])

                if mem[j][k] > sol:
                    sol = mem[j][k]
                    sol_values = s[k:j+1]
        
        return sol_values

File: 5-longest-palindromic-substring/longest-palindromic-substring.cpp
struct sol {
    int size;
    int i;
    int j;

    sol(const int size, const int i, const int j): size(size), i(i), j(j) {}

    auto operator<=>(const sol& other) const {
        return size <=> other.size;
    }

    bool operator==(const sol& other) const {
        return size == other.size;
    }
};

class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        sol ans = sol(1,0,0);

        vector<vector<bool>> dp(n, vector<bool> (n, true));

        for(int i = n-2; i >= 0; i--){
            for(int j = n-1; j>i ; j--){
                dp[i][j] = s[i] == s[j] && dp[i+1][j-1];
                if(dp[i][j]){
                    ans = max(ans, sol(j-i+1, i, j));
                }
            }
        }
        return s.substr(ans.i, ans.size);
    }
};